### PR TITLE
Fix Add support for multi-events on Billetweb #3

### PR DESCRIPTION
**Pull request by csauge**
- **Increase the number of workshops listed by 2** (from 262 to 537)
- It reduces the **processing time by 50%** for billetweb (from 16 to 8 minutes)
- Add +171 lines and remove −209,723 lines (mainly due to json removal)
- Remove multi and filter tests because seems useless
- Suppress useless geopy in requirements.txt
- Suppress useless imports, comments, error managements
- Simplify new_ui logic
- Simplify readJson.py to get the departement from API directly and reduce **processing time by 10%** for global run
- Remove useless data directory with the two json
- Simplify and robustify the event_date parsing with regex
- Fix bug with "at" dates wrong using a previous event end_date. Use 3 hours default duration.
- Add checks to ensure that all pages/iframe readyState are completed
- uuids for multi-time include the session_id
- Fix gift card for " don" instead of "don"
- Robustify and simplify sold_out
- Increase Rejecting print error displays
- Add more cases for training = true

### CODE REVIEW

**Comment by csauge** [comment]
not sure about "The WebDriverWait robustness helps to capture +20% results in json (4300 lines to 5600 lines)".
It seems that the results are not deterministics.
But this is at least as good :)

**Comment by csauge** [comment]
I think you can try to deliver this (after your test of course), because my next step would be to manage the calendar in billetweb (which seems not easy)

**Review by thomas-bouvier — CHANGES_REQUESTED** [review]
Merci pour ton travail @csauge. On gagne en performance et en exhaustivité, ce qui est génial. On peut discuter sous mes commentaires !

File: scraper/billetweb.py [hunk 1]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
-                print(f"\n-> Processing {link} ...")
-                driver.get(link)
-                driver.implicitly_wait(3)
-
-                if "filter" in page:
-                    if page["filter"] not in link:
-                        print(
-                            "Rejecting filter: expected filtering keyword not present in current link"
-                        )
-                        continue
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:144]: OK

File: scraper/billetweb.py [hunk 2]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
+            # Description
+            try:
+                driver.find_element(By.ID, "more_info").click()
+            except Exception:
+                pass  # normal case if description is small or old ui
+            even_el = driver.find_element(by=By.CSS_SELECTOR, value="#description")
+            description = even_el.text
+
+            # Parse event id
+            event_id = re.search(r"/([^/]+?)&", link).group(1)
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:147]: OK

File: scraper/billetweb.py [hunk 3]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
+            driver.get(link)
+            wait.until(lambda driver: driver.execute_script('return document.readyState') == 'complete')
+
+            # Description
+            try:
+                driver.find_element(By.ID, "more_info").click()
+            except Exception:
+                pass  # normal case if description is small or old ui
+            even_el = driver.find_element(by=By.CSS_SELECTOR, value="#description")
+            description = even_el.text
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:144]: OK

File: scraper/billetweb.py [hunk 4]
```diff
@@ -349,83 +338,29 @@ def get_billetweb_data(service, options):
...
-                    ############################################################
-                    try:
-                        search_query = f"{address}, {city}, France"
-                        address_dict = get_address_data(search_query)
-                    except json.JSONDecodeError:
-                        print(
-                            "Rejecting record: error while parsing the national address API response"
-                        )
-                        continue
-
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:363]: OK

File: utils/readJson.py [hunk 5]
```diff
@@ -1,89 +1,42 @@
...
         longitude = data["features"][0]["geometry"]["coordinates"][0]
-    except:
+    except Exception:
         longitude = ""
     try:
         latitude = data["features"][0]["geometry"]["coordinates"][1]
-    except:
+    except Exception:
         latitude = ""
-    cod_dep = department_code(postcode)
```
[the remark this answers is not in the document]
Comment (csauge) [reply utils/readJson.py:75]: Comme convenu on le garde dans le PR

File: utils/readJson.py [hunk 6]
```diff
@@ -1,89 +1,42 @@
...
+    try:
+        cod_dep = data["features"][0]["properties"]["context"].split(',')[0].strip()
```
[the remark this answers is not in the document]
Comment (csauge) [reply utils/readJson.py:28]: Comme convenu on le garde dans le PR

File: utils/readJson.py [hunk 7]
```diff
@@ -1,89 +1,42 @@
...
-    }
-
-    try:
-        response = requests.request(
-            "GET", url, data=payload, headers=headers, params=querystring
-        )
-        data = response.json()
-    except json.JSONDecodeError as e:
-        raise e
+    params = {"q": text_address, "limit": 1}
```
[the remark this answers is not in the document]
Comment (csauge) [reply utils/readJson.py:7]: Permet de garder seulement le premier résultat (suffisant aujourd'hui)

File: scraper/billetweb.py [hunk 8]
```diff
@@ -349,83 +338,29 @@ def get_billetweb_data(service, options):
...
-                        By.CSS_SELECTOR,
-                        "div.block",
-                    )
-                    sold_out = (
-                        "aucune" in remaining_slots_el.text
-                        or "nombre maximal" in remaining_slots_el.text
-                        or "sold out" in remaining_slots_el.text
-                    )
-                except NoSuchElementException:
-                    sold_out = "complet" in title.lower()
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:414]: en fait l'existance meme de div.block suffit à dire que c'est sold out (full or waiting list). seul un cas (géré) permet de dire que ce n'est pas sold out si l'evenement est à payer par un autre moyen (par exemple vers un autre site)

File: data/codepostal_to_insee.json
[the remark this answers is not in the document]
Comment (csauge) [reply data/codepostal_to_insee.json:1]: Comme convenu on le garde dans le PR

File: data/insee_to_codepostal_and_code_departement.json
[the remark this answers is not in the document]
Comment (csauge) [reply data/insee_to_codepostal_and_code_departement.json:1]: Comme convenu on le garde dans le PR

File: scraper/billetweb.py [hunk 9]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
+                finally:
+                    driver.switch_to.parent_frame()
+        
+                event_info.append([main_title, event_time, main_full_location, sold_out, link, event_id])
+
+            ################################################################
+            # Event loop
+            ################################################################
+            for title, event_time, full_location, sold_out, ticket_link, uuid in event_info:
+                print(f"\n-> Processing {title}, {event_time}, {full_location}, {sold_out}, {ticket_link}, {uuid} ...")
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:255]: OK

File: scraper/billetweb.py [hunk 10]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
+                    event_start_datetime = parse(f"{match.group('date')} {match.group('start')}")
+                    event_end_datetime = parse(f"{match.group('date')} {match.group('end')}")
+                elif match := re.match(r"(?P<start_date>.*)\sat\s(?P<start_time>.*)\sto\s(?P<end_date>.*)\sat\s(?P<end_time>.*)", event_time):
+                    # Thu Oct 19, 2023 at 01:00 PM to Sat Feb 24, 2024 at 02:00 PM
+                    event_start_datetime = parse(f"{match.group('start_date')} {match.group('start_time')}")
+                    event_end_datetime = parse(f"{match.group('end_date')} {match.group('end_time')}")
+                elif match := re.match(r"(?P<date>.*)\sat\s(?P<time>.*)", event_time):
+                    # Thu Oct 19, 2023 at 01:00 PM
+                    event_start_datetime = parse(f"{match.group('date')} {match.group('time')}")
+                    event_end_datetime = event_start_datetime + timedelta(hours=3)
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:271]: Oui tous les cas sont bien gérés

File: scraper/billetweb.py [hunk 11]
```diff
@@ -123,176 +125,163 @@ def get_billetweb_data(service, options):
...
+                main_full_location = ""
+
+            event_info = []
+
+            ################################################################
+            # Multi-time management
+            ################################################################
+            wait.until(EC.frame_to_be_available_and_switch_to_it((By.CSS_SELECTOR, "#shop_block iframe")))
+            wait.until(lambda driver: driver.execute_script('return document.readyState') == 'complete')
+            sessions_ele = driver.find_elements(By.CSS_SELECTOR, 'a.sesssion_href')
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:181]: Bizarre mais oui :)

File: scraper/billetweb.py [hunk 12]
```diff
@@ -302,29 +291,29 @@ def get_billetweb_data(service, options):
...
+                    # Rejecting record: empty address: Espace Multiburo - 1 Cr du Havre 75008 PARIS
+                    # Rejecting record: empty address: Online
+                    # Rejecting records: address is too long (5 parts): La Melting Coop, épicerie et café participatifs, Cours Emile Zola, Villeurbanne, Fr
+                    # Rejecting record: empty address: Salle des fêtes de Ramonville Rue Joliot Cur
+                    # Rejecting record: empty address: Le LAC de Sèvres 12 rue Lecointre 92310 Sèvres
+                    # Rejecting record: empty address: La BOM 2 rue Girard 93100 Montreuil
+                    # Rejecting record: empty address: Les Canaux 6 quai de la Seine - 75019 Paris
+                    # Rejecting records: address is too long (5 parts): Canopy, l'art de se réunir!, Rue Négrier, Lille, France
+                    # A FAIRE: supprimer les noms de villes non francaises
+                    # FEATURES TO ADD: manage outside france, timezones, number of remainings, price
```
[the remark this answers is not in the document]
Comment (csauge) [reply scraper/billetweb.py:310]: OK. I have created 2 PR for this comments to manage

**Comment by thomas-bouvier** [comment]
Ferme #3. Merci beaucoup pour ton travail @csauge, nous passons de ~ateliers en présentiel recensés à presque 800.

**Pull request merged**